Add documentation for creating custom rules. Closes #261

floroz opened the pull request:
This PR adds a a step by step guide on how to create and add a custom rule to `remark-lint`.

The newly created guide is also added as a new section in the `readme.md`.

- Closes https://github.com/remarkjs/remark-lint/issues/261

floroz committed:
chore: add documentation for custom rules

floroz committed:
extend example to include config from start to finish

floroz committed:
add table of contents

floroz committed:
add table of contents

floroz committed:
toc bug

floroz commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,211 @@
+*   [Import the rule in your remark config](#import-the-rule-in-your-remark-config)
+*   [Apply the rule](#apply-the-rule)
+
+\## Set up the project
```
@wooorm 
not sure why I get this autofix for this line:

```bash

doc/create-a-custom-rule.md
  7:5-7:46  warning  Link to unknown heading: `set-up-the-project`  missing-heading  remark-validate-links
```

it only happens for this first header.

floroz committed:
readme update

Commit:
updated custom rule

Commit:
fix bash commands

Commit:
fix bash commands

Commit:
fix broken space

Commit:
add bash command

Commit:
update intro

floroz commented:
@wooorm cannot add reviewers but this PR is ready.

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+This guide is part of [a step-by-step tutorial](https://dev.to/floroz/how-to-create-a-custom-lint-rule-for-markdown-and-mdx-using-remark-and-eslint-2jim), and will help you getting started to create your first linting plugin for `remark`.
+
+## Table of Contents
```
```suggestion
## Contents
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## Set up the project
+
+Create a new folder and navigate inside it from your Terminal. For this example I will be using UNIX commands (macOS and Linux compatible).
+Now we can generate our `package.json`
```
```suggestion
Create a new folder and enter it from your terminal. For this example I will be using Unix commands (macOS and Linux compatible).
Now we can generate our `package.json`
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+## Set up remark
+
+With our dependencies all installed, we can start creating a `.remarkrc.js`, which will contain all the plugins that will be consumed by the remark processor.
+For details about alternative or advanced configurations, please refer to [Configuring remark-lint](https://github.com/remarkjs/remark-lint#configuring-remark-lint).
```
```suggestion
For more info on configuration, see [Configuring `remark-lint`](https://github.com/remarkjs/remark-lint#configuring-remark-lint).
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+Now we can generate our `package.json`
+
+```bash
+$ mkdir my-custom-rule
+
+$ cd my-custom-rule
+
+$ yarn init -y
````
```suggestion
mkdir my-custom-rule
cd my-custom-rule
npm init -y
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+Now we can start installing our dependencies.
+
+```bash
+$ yarn add remark-lint remark-cli
````
```suggestion
npm install remark-lint remark-cli -D
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+$ yarn add remark-lint remark-cli
+```
+
+*   [remark-lint](https://github.com/remarkjs/remark-lint): a plugin to lint markdown built on [remark](https://github.com/remarkjs/remark): (a markdown processor).
+*   [remark-cli](https://github.com/remarkjs/remark/tree/main/packages/remark-cli): remark CLI.
````
```suggestion
*   [`remark-lint`](https://github.com/remarkjs/remark-lint): Core lint plugin
*   [`remark-cli`](https://github.com/remarkjs/remark/tree/main/packages/remark-cli): Command-line interface
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+*   [remark-lint](https://github.com/remarkjs/remark-lint): a plugin to lint markdown built on [remark](https://github.com/remarkjs/remark): (a markdown processor).
+*   [remark-cli](https://github.com/remarkjs/remark/tree/main/packages/remark-cli): remark CLI.
+
+Because we will be working with [ASTs](https://en.wikipedia.org/wiki/Abstract_syntax_tree), we will also need some utilities:
```
```suggestion
We will also need some utilities:
```

The utilities are not necessarily needed for ASTs, so I suggest removing that here. You could add it somewhere else.

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## Set up remark
+
+With our dependencies all installed, we can start creating a `.remarkrc.js`, which will contain all the plugins that will be consumed by the remark processor.
```
```suggestion
With everything installed, we can now create a `.remarkrc.js` that will contain the plugins we’ll use.
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+}
+```
+
+Let's create a `doc.md`, the markdown file we want to lint,
````
```suggestion
Let's create a `doc.md`, the markdown file we want to lint:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+$ touch doc.md
+```
+
+and copy/paste this content:
````
```suggestion
…and copy/paste the following:
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+At this point, we have a working `remark` configuration and a markdown file in the project.
+
+If we run `yarn run lint` we should expect to see in our terminal:
```
```suggestion
Now running `npm run lint` should show:
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+[Back to Top](#table-of-contents)
+
+## The no-invalid-gif rule
```
```suggestion
## The `no-invalid-gif` rule
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## The no-invalid-gif rule
+
+Let's imagine we want to write a rule that checks whether a `.gif` file is used within an image.
```
```suggestion
Let’s imagine we want to write a rule that checks whether a `.gif` file is used as an image.
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+![a funny cat](funny-cat.gif)
+```
+
+Because the file extension `.gif` in the image tag, violates our rule.
````
```suggestion
Because the file extension `.gif` in the image tag violates our rule.
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+$ cd .. # return to project root
+```
+
+*Remember*: the name of the folder and files, and where to place them within your project, is entirely up to you.
````
```suggestion
*Note*: the name of folders and files, and where to place them within your project, is up to you.
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+*Remember*: the name of the folder and files, and where to place them within your project, is entirely up to you.
+
+In `./rules/no-gif-allowed.js`, let's import the `unified-lint-rule`.
```
```suggestion
In `./rules/no-gif-allowed.js`, let's import `unified-lint-rule`.
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+module.exports = rule("my-npm-published-package:no-gif-allowed", noGifAllowed);
+```
+
+This can help you when wanting to create a group of rules under the same *label* or *namespace*.
````
```suggestion
This can help you when wanting to create a group of rules under the same *namespace*.
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+function noGifAllowed(tree, file, options) {}
+```
+
+*   `tree` (*required*): a [mdast](https://github.com/syntax-tree/mdast).
+*   `file` (*required*): a [virtual file format](https://github.com/vfile/vfile).
````
```suggestion
*   `tree` (*required*): [mdast](https://github.com/syntax-tree/mdast)
*   `file` (*required*): [virtual file](https://github.com/vfile/vfile)
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+*   `tree` (*required*): a [mdast](https://github.com/syntax-tree/mdast).
+*   `file` (*required*): a [virtual file format](https://github.com/vfile/vfile).
+*   `options` (*optional*): additional information passed to the rule by the remark plugins definition.
```
```suggestion
*   `options` (*optional*): additional information passed to the rule by users
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## Rule implementation
+
+Because we will be inspecting a [mdast](https://github.com/syntax-tree/mdast), which is a markdown abstract syntax tree built upon [unist](https://github.com/syntax-tree/unist), we can take advantage of the many existing [unist utilities](https://github.com/syntax-tree/unist#utilities) to inspect our tree's nodes.
```
```suggestion
Because we will be inspecting [mdast](https://github.com/syntax-tree/mdast), which is a markdown abstract syntax tree built upon [unist](https://github.com/syntax-tree/unist), we can take advantage of the many existing [unist utilities](https://github.com/syntax-tree/unist#utilities) to inspect our tree’s nodes.
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+Now that our custom rule is defined, and ready to be used, we need to add it to our `remark` configuration.
+
+All you have to do is to import your rule into the `remark` configuration plugins array:
```
```suggestion
You can do that by importing your rule and adding it in `plugins` array:
```

wooorm reviewed (commented):
Nice work! Here are a bunch of editorial suggestions. I’ll do a more content-focussed review next.

* Could you use ` ```sh ` for bash examples?
* Could you use ` ```markdown ` for markdown examples (I believe GH didn’t highlight `md`)?
* Could you remove `$` from bash examples (when they don’t show output too). That dollar makes it hard to copy
* Could you match code-style to the one used in this project? Especially consistency with trailing commas and quote use
* Similarly, I see both `//` and `/*` comment styles, sometimes capitalized, sometimes not. Pick one, preferably capitalized.
* Could you use “smart” (curly) apostrophes? `’`, not `'`
* Can you put enters after sentences? (helps with git diffs)
* I quite prefer reference style links `[text][id]` in prose and `[id]: url` at the bottom, can I persuade you to do that to?
* I’d prefer using `npm` in examples over `yarn`, as I believe npm is still more common thus match mores folks’ workflows

---

Here are some tools that can help you write better: https://github.com/get-alex/alex, https://wooorm.com/write-music/, https://wooorm.com/readability/.

You might also avoid [hedges](https://en.wikipedia.org/wiki/Hedge_(linguistics)) (https://github.com/words/hedges), [fillers](https://en.wikipedia.org/wiki/Filler_(linguistics)) (https://github.com/words/fillers), and [weasels](https://en.wikipedia.org/wiki/Weasel_word) (https://github.com/words/weasels).

Commit:
address PR comments

floroz commented:
> Nice work! Here are a bunch of editorial suggestions. I’ll do a more content-focussed review next.
> 
> * Could you use `` ```sh `` for bash examples?
> * Could you use `` ```markdown `` for markdown examples (I believe GH didn’t highlight `md`)?
> * Could you remove `$` from bash examples (when they don’t show output too). That dollar makes it hard to copy
> * Could you match code-style to the one used in this project? Especially consistency with trailing commas and quote use
> * Similarly, I see both `//` and `/*` comment styles, sometimes capitalized, sometimes not. Pick one, preferably capitalized.
> * Could you use “smart” (curly) apostrophes? `’`, not `'`
> * Can you put enters after sentences? (helps with git diffs)
> * I quite prefer reference style links `[text][id]` in prose and `[id]: url` at the bottom, can I persuade you to do that to?
> * I’d prefer using `npm` in examples over `yarn`, as I believe npm is still more common thus match mores folks’ workflows
> 
> Here are some tools that can help you write better: https://github.com/get-alex/alex, https://wooorm.com/write-music/, https://wooorm.com/readability/.
> 
> You might also avoid [hedges](https://en.wikipedia.org/wiki/Hedge_(linguistics)) (https://github.com/words/hedges), [fillers](https://en.wikipedia.org/wiki/Filler_(linguistics)) (https://github.com/words/fillers), and [weasels](https://en.wikipedia.org/wiki/Weasel_word) (https://github.com/words/weasels).

That's some great feedback, thank you!

I have addressed your comments, I am just not sure about:

> * I quite prefer reference style links `[text][id]` in prose and `[id]: url` at the bottom, can I persuade you to do that to?

Would you mind help me understand more about this?

Commit:
comment capital letter

Commit:
lint fix

wooorm commented:
> Would you mind help me understand more about this?

See the alternative version of links here: https://commonmark.org/help/. And see examples in the readme: https://raw.githubusercontent.com/remarkjs/remark-lint/main/readme.md

floroz commented:
> > Would you mind help me understand more about this?
> 
> See the alternative version of links here: https://commonmark.org/help/. And see examples in the readme: https://raw.githubusercontent.com/remarkjs/remark-lint/main/readme.md

Got it! didn't know that alternative syntax.

If that's okay with you I would prefer leaving them as they are, ultimately they live in a separate file so they won't be creating any confusion with the other `readme.md` docs.

Besides that, everything else should have been fixed, let me know if you are happy to merge it!

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+Now we can generate our `package.json`
+
+```sh
+ mkdir my-custom-rule
+
+ cd my-custom-rule
+
+ npm init -y
````
```suggestion
mkdir my-custom-rule
cd my-custom-rule
npm init -y
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+## Set up the project
+
+Create a new folder and enter it from your terminal. For this example I will be using Unix commands (macOS and Linux compatible).
+Now we can generate our `package.json`
```
```suggestion
Then generate a `package.json`:
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## Set up the project
+
+Create a new folder and enter it from your terminal. For this example I will be using Unix commands (macOS and Linux compatible).
```
```suggestion
Create a new folder and enter it from your terminal.
For this example I will be using Unix commands (macOS and Linux compatible).
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+ npm init -y
+```
+
+Now we can start installing our dependencies.
````
```suggestion
Now we can start installing our dependencies:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+Now we can start installing our dependencies.
+
+```sh
+ npm install remark-lint remark-cli
````
```suggestion
npm install remark-lint remark-cli
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+We will also need some utilities:
+
+```sh
+ npm install unified-lint-rule unist-util-generated unist-util-visit
````
```suggestion
npm install unified-lint-rule unist-util-generated unist-util-visit
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+*   [`remark-lint`](https://github.com/remarkjs/remark-lint): Core lint plugin
+*   [`remark-cli`](https://github.com/remarkjs/remark/tree/main/packages/remark-cli): Command-line interface
+
+We will also need some utilities:
```
```suggestion
We will also use some utilities:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+For more info on configuration, see [Configuring `remark-lint`](https://github.com/remarkjs/remark-lint#configuring-remark-lint).
+
+```sh
+ touch .remarkrc.js
````
```suggestion
touch .remarkrc.js
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+
+```js
+// .remarkrc.js
+
+module.exports = {
+  plugins: [],
+};
````
```suggestion
module.exports = {
  plugins: []
}
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+};
+```
+
+Then, in our `package.json`, let's add the following script, which will process all the markdown file within our project:
````
```suggestion
Then, in our `package.json`, add the following script to process all the markdown files in our project:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+}
+```
+
+Let's create a `doc.md`, the markdown file we want to lint:
````
```suggestion
Let’s create a `doc.md`, the markdown file we want to lint:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+Let's create a `doc.md`, the markdown file we want to lint:
+
+```sh
+ touch doc.md
````
```suggestion
touch doc.md
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+ touch doc.md
+```
+
+...and copy/paste the following:
````
```suggestion
…and copy/paste the following:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+If we run `npm run lint` we should expect to see in our terminal:
+
+```sh
+ doc.md: no issues found
````
```suggestion
doc.md: no issues found
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+ doc.md: no issues found
+```
+
+All good, the file has been processed, and because we haven't specified any plugins nor lint rule, no issues are found.
````
```suggestion
All good, the file has been processed, and because we haven’t specified any plugins nor lint rules, no issues are found.
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## The `no-invalid-gif` rule
+
+Let’s imagine we want to write a rule that checks whether a `.gif` file is used as an image.
+
+Given the content of our `doc.md` file declared above, we would expect an *error* or *warning* pointing to:
```
```suggestion
Let’s imagine we want to write a rule that checks whether a `.gif` file is used as an image.
Given the content of our `doc.md` file declared above, we would expect an *error* or *warning* pointing to:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+![a funny cat](funny-cat.gif)
+```
+
+Because the file extension `.gif` in the image tag violates our rule.
````
```suggestion
Because the file extension `.gif` in the image violates our rule.
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## Create the custom rule
+
+Let's create a new folder `rules` under the root directory, where we will place all of our custom rules, and create a new file in it named `no-gif-allowed.js`.
```
```suggestion
Let’s create a new folder `rules` under the root directory, where we will place all of our custom rules, and create a new file in it named `no-gif-allowed.js`.
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+Let's create a new folder `rules` under the root directory, where we will place all of our custom rules, and create a new file in it named `no-gif-allowed.js`.
+
+```sh
+ mkdir rules
+ cd rules
+ touch no-gif-allowed.js
+ cd .. # return to project root
````
```suggestion
mkdir rules
cd rules
touch no-gif-allowed.js
cd .. # return to project root
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+*Note*: the name of folders and files, and where to place them within your project, is up to you.
+
+In `./rules/no-gif-allowed.js`, let's import `unified-lint-rule`.
```
```suggestion
In `./rules/no-gif-allowed.js`, let’s import `unified-lint-rule`.
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+```js
+// rules/no-gif-allowed.js
+
+var rule = require("unified-lint-rule");
+function noGifAllowed(tree, file, options) {
+  // rule implementation
+}
+module.exports = rule("remark-lint:no-gif-allowed", noGifAllowed);
````
```suggestion
var rule = require('unified-lint-rule')
function noGifAllowed(tree, file, options) {
  // Rule implementation
}
module.exports = rule('remark-lint:no-gif-allowed', noGifAllowed)
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+module.exports = rule("remark-lint:no-gif-allowed", noGifAllowed);
+```
+
+Let's say you want all your custom rules to be defined as part of your project namespace. If your project was named `my-project`, then you can export your rule as:
````
```suggestion
Let’s say you want all your custom rules to be defined as part of your project namespace.
If your project was named `my-project`, then you can export your rule as:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+Let's say you want all your custom rules to be defined as part of your project namespace. If your project was named `my-project`, then you can export your rule as:
+
+```js
+module.exports = rule("my-project-name:no-gif-allowed", noGifAllowed);
+// or
+module.exports = rule("my-npm-published-package:no-gif-allowed", noGifAllowed);
````
```suggestion
module.exports = rule('my-project-name:no-gif-allowed', noGifAllowed)
// Or:
module.exports = rule('my-npm-published-package:no-gif-allowed', noGifAllowed)
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## Rule arguments
+
+Your rule function will receive three arguments.
```
```suggestion
Your rule function will receive three arguments:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+For this example, we will use [`unist-util-visit`](https://github.com/syntax-tree/unist-util-visit) to recursively inspect all the image nodes, and [`unist-util-generated`](https://github.com/syntax-tree/unist-util-generated) to ensure we are not inspecting nodes that we have generated ourselves and do not belong to the `doc.md`.
+
+```js
+const rule = require("unified-lint-rule");
+const visit = require("unist-visit-util");
+const generated = require("unist-util-generated");
````
```suggestion
const rule = require('unified-lint-rule')
const visit = require('unist-visit-util')
const generated = require('unist-util-generated')
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+  // Here we check whether the given node violates our rule.
+  // Implementation details are not relevant to the scope of this example.
+  // This is an overly simplified solution for demonstration purposes
+  if (node.url && typeof node.url === "string") {
+    return !node.url.endsWith(".gif");
```
```suggestion
  if (node.url && typeof node.url === 'string') {
    return !node.url.endsWith('.gif')
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+  }
+}
+function noGifAllowed(tree, file, options) {
+  visit(tree, "image", visitor);
```
```suggestion
  visit(tree, 'image', visitor)
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+       // This is an extremely simplified example of how to structure
+       // the logic to check whether a node violates your rule.
+       // You have complete freedom over how to visit/inspect the tree,
+       //and on how to implement the validation logic for your node.
+      const isValid = isValidNode(node);
```
```suggestion
       // and on how to implement the validation logic for your node.
      const isValid = isValidNode(node)
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+        // Remember to provide the node as second argument to the message,
+        // in order to obtain the position and column where the violation occurred.
+        file.message(
+          `Invalid image file extentions. Please do not use gifs`,
+          node
+        );
```
```suggestion
          'Invalid image file extentions. Please do not use gifs',
          node
        )
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+    }
+  }
+}
+module.exports = rule("remark-lint:no-gif-allowed", noGifAllowed);
```
```suggestion
module.exports = rule('remark-lint:no-gif-allowed', noGifAllowed)
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## Import the rule in your remark config
+
+Now that our custom rule is defined, and ready to be used, we need to add it to our `remark` configuration.
```
```suggestion
Now that our custom rule is defined and ready to be used we need to add it to our `remark` configuration.
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+
+```js
+// .remarkrc.js
+const noGifAllowed = require("./rules/no-gif-allowed.js");
+
+module.exports = {
+  plugins: [noGifAllowed],
+};
````
```suggestion
const noGifAllowed = require('./rules/no-gif-allowed.js')

module.exports = {
  plugins: [noGifAllowed]
}
```

wooorm commented on doc/create-a-custom-rule.md:
```diff
@@ -0,0 +1,242 @@
+
+## Apply the rule on the Markdown file
+
+If you run `npm lint`, you should see the following message in the terminal:
```
```suggestion
If you run `npm run lint`, you should see the following message in the terminal:
```

wooorm commented on doc/create-a-custom-rule.md:
````diff
@@ -0,0 +1,242 @@
+
+If you run `npm lint`, you should see the following message in the terminal:
+
+```sh
+ 5:1-5:30  warning  Invalid image file extentions. Please do not use gifs  no-gif-allowed  remark-lint
````
````suggestion
```text
5:1-5:30  warning  Invalid image file extentions. Please do not use gifs  no-gif-allowed  remark-lint
````

wooorm reviewed (changes requested):
Nice! Some more nits. (btw: did you know you can commit these suggestions through the GH UI? Just checking, that’ll help speed things up).
Other than these nits the article reads great!

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz committed:
Update doc/create-a-custom-rule.md

Co-authored-by: Titus <tituswormer@gmail.com>

floroz commented:
> Nice! Some more nits. (btw: did you know you can commit these suggestions through the GH UI? Just checking, that’ll help speed things up).
> Other than these nits the article reads great!

Thanks for taking the time to review in such details!

Yes I am aware, I generally review/work on PR directly within VS Code using the [GitHub PR Extension](https://code.visualstudio.com/blogs/2018/09/10/introducing-github-pullrequests), but considering the number of changes request this time I used the GH UI. 
I am curious though, was there a way for me to accept all suggestions at once in a single commit?

ChristianMurphy commented:
> I am curious though, was there a way for me to accept all suggestions at once in a single commit?

In the "files changed" tab, it should offer to add changes to a batch, which will be a single commit.
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request

floroz commented:
> > I am curious though, was there a way for me to accept all suggestions at once in a single commit?
> 
> In the "files changed" tab, it should offer to add changes to a batch, which will be a single commit.
> https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request

Thank you @ChristianMurphy ! That's really handy to know 😄 sorry about the commit spam then!

wooorm commented:
Awesome!

wooorm commented:
Thanks for all your work on this, @floroz!

floroz commented:
> Thanks for all your work on this, @floroz!

Thanks for the feedback, happy to help!
